### ISSUE TITLE
VAE Decoder Usage

### ISSUE THREAD

**Issue by qiqiApink** [Yikai-Wang/asuka-misato#11]
Can this vae decoder work with diffusers fluxfill pipeline?

**Comment by Yikai-Wang** [comment]
Sure. You can set the pipeline's output mode to 'latent' directly, then use our decode function:
https://github.com/Yikai-Wang/asuka-misato/blob/35c65705b54034be389a6cdcc9f29b5c9452e978/asuka-flux/test_asuka_flux.py#L96
Alternatively, you can modify the relevant decoding code in the diffusers pipeline:
https://github.com/huggingface/diffusers/blob/754fe85cace17ae8615d53578d6d842c9e4a1bd9/src/diffusers/pipelines/flux/pipeline_flux_fill.py#L1021-L1029

**Comment by qiqiApink** [comment]
> Sure. You can set the pipeline's output mode to 'latent' directly, then use our decode function:
> 
> [asuka-misato/asuka-flux/test_asuka_flux.py](https://github.com/Yikai-Wang/asuka-misato/blob/35c65705b54034be389a6cdcc9f29b5c9452e978/asuka-flux/test_asuka_flux.py#L96)
> 
> Line 96 in [35c6570](/Yikai-Wang/asuka-misato/commit/35c65705b54034be389a6cdcc9f29b5c9452e978)
> 
>  preds = vae.my_decode(denoised_x, orig_img, mask) 
> 
> Alternatively, you can modify the relevant decoding code in the diffusers pipeline:
> https://github.com/huggingface/diffusers/blob/754fe85cace17ae8615d53578d6d842c9e4a1bd9/src/diffusers/pipelines/flux/pipeline_flux_fill.py#L1021-L1029

I'm having trouble with batch inference. When each sample is decoded separately, everything looks fine. However, when the batch_size is 2, one of the samples cannot be decoded correctly as follows

<img width="768" height="768" alt="Image" src="https://github.com/user-attachments/assets/ae298b74-a8db-4a2c-8f08-fd6e1f6b9c00" />

**Comment by Yikai-Wang** [comment]
It looks like there's something wrong with your batch implementation. You might double-check how the batch channel and repeat/clone operators are used in each variable to check if they work as expected.

**Comment by qiqiApink** [comment]
> It looks like there's something wrong with your batch implementation. You might double-check how the batch channel and repeat/clone operators are used in each variable to check if they work as expected.

Well, I'm sure the code is exactly right. I just replaced the original vae decoder with yours and the original vae decoder works fine.

**Comment by Yikai-Wang** [comment]
Have you tried running test_asuka_flux.py? It also performs batch inference—does it show the same issue?
Also, could you share an example script so we can check whether we can reproduce and fix the problem?

**Comment by qiqiApink** [comment]
> Have you tried running test_asuka_flux.py? It also performs batch inference—does it show the same issue? Also, could you share an example script so we can check whether we can reproduce and fix the problem?

I haven't tried it yet because I want to use the vae decoder only. My test script is as follows:
```
from omegaconf import OmegaConf
from ldm.modules.diffusionmodules.asuka_decoder import Decoder
from einops import rearrange
ddconfig = OmegaConf.load("./configs/condition_decoder.yaml")
decoder = Decoder(**ddconfig)
state_dict = torch.load('ASUKA-FLUX.1-Fill/asuka_decoder.ckpt', map_location=torch.device('cpu'))['state_dict']
flux_ae = {}
for k, v in state_dict.items():
    if k.startswith("encoder.") or k.startswith("decoder."):
        flux_ae[k] = v

pipe = FluxFillPipeline.from_pretrained(
    'FLUX.1-Fill-dev',
    torch_dtype=torch.bfloat16,
)
decoder = decoder.to(dtype=torch.bfloat16)
pipe.vae.decoder = decoder
pipe.vae.load_state_dict(flux_ae, strict=False)
pipeline = pipe.to('cuda')

inp_imgs = [Image.open('crop1.png'), Image.open('crop2.png')]
inp_masks = [Image.open('mask1.png').convert('L'), Image.open('mask2.png').convert('L')]
imgs = torch.from_numpy(np.stack(inp_imgs, axis=0)) / 127.5 - 1.0
imgs = rearrange(imgs, "b h w c -> b c h w").to(torch.bfloat16).to('cuda')
masks = np.stack(inp_masks, axis=0) / 255.
masks = torch.from_numpy(masks[:, None]).to(torch.bfloat16).to('cuda')

validation_prompt = ["empty scene"] * len(inp_imgs)
prompt_embeds, pooled_prompt_embeds, text_ids = pipeline.encode_prompt(
    validation_prompt, prompt_2=validation_prompt
)

with torch.no_grad():
    latents = pipeline(
        prompt_embeds=prompt_embeds,
        pooled_prompt_embeds=pooled_prompt_embeds,
        image=inp_imgs,
        mask_image=inp_masks,
        num_inference_steps=20,
        height=768,
        width=768,
        guidance_scale=50,
        generator=torch.Generator('cuda').manual_seed(42),
        output_type='latent',
    ).images
    latents = pipeline._unpack_latents(latents, 768, 768, pipeline.vae_scale_factor)
    latents = (latents / pipeline.vae.config.scaling_factor) + pipeline.vae.config.shift_factor
    images = pipeline.vae.decoder(latents.to(torch.bfloat16).to('cuda'), imgs, masks)
    images = pipeline.image_processor.postprocess(images, output_type='pil')
    images[0].save('output1.png')
    images[1].save('output2.png')
```

**Comment by qiqiApink** [comment]
Btw, If I replace
```
images = pipeline.vae.decoder(latents.to(torch.bfloat16).to('cuda'), imgs, masks)
images = pipeline.image_processor.postprocess(images, output_type='pil')
```
with
```
images1 = pipeline.vae.decoder(latents[0:1].to(torch.bfloat16).to('cuda'), imgs[0:1], masks[0:1])
images2 = pipeline.vae.decoder(latents[1:2].to(torch.bfloat16).to('cuda'), imgs[1:2], masks[1:2])
image1 = pipeline.image_processor.postprocess(images1, output_type='pil')[0]
image2 = pipeline.image_processor.postprocess(images2, output_type='pil')[0]
```
The results will be right.

**Comment by Yikai-Wang** [comment]
Hi there,
We've identified the cause of the issue and updated our code to fix the bug. The problem was due to the interpolation method used here:
https://github.com/Yikai-Wang/asuka-misato/blob/1a6b6817f1d700d481bbc7ca48ff1d2b3674c3fd/asuka-flux/ldm/modules/diffusionmodules/asuka_decoder.py#L446
It turns out that this method can be unstable in your use case, even though it works fine on our end. We’ve replaced it with a manual  one-by-one interpolation. With this change, the decoding results should now be correct. Please give it a try and let us know if you run into any further issues.

**Comment by qiqiApink** [comment]
> Hi there, We've identified the cause of the issue and updated our code to fix the bug. The problem was due to the interpolation method used here:
> 
> [asuka-misato/asuka-flux/ldm/modules/diffusionmodules/asuka_decoder.py](https://github.com/Yikai-Wang/asuka-misato/blob/1a6b6817f1d700d481bbc7ca48ff1d2b3674c3fd/asuka-flux/ldm/modules/diffusionmodules/asuka_decoder.py#L446)
> 
> Line 446 in [1a6b681](/Yikai-Wang/asuka-misato/commit/1a6b6817f1d700d481bbc7ca48ff1d2b3674c3fd)
> 
>  F.interpolate(mask[i:i+1], size=h.shape[-2:], mode='area') for i in range(mask.shape[0]) 
> 
> It turns out that this method can be unstable in your use case, even though it works fine on our end. We’ve replaced it with a manual one-by-one interpolation. With this change, the decoding results should now be correct. Please give it a try and let us know if you run into any further issues.

Excellent! Now, the results look great. However, I'm still confused about this. The interpolation operation doesn't process the first dimension (batch). In my opinion, there is no difference between these two formulations.

**Comment by Yikai-Wang** [comment]
We're also confused by this. It works normally on our end, but not in your case. We haven't been able to pinpoint the cause — it might be an internal bug triggered by different ways of importing the decoder, but we couldn't identify the exact reason. The good thing is, at least we now know how to avoid it.😂
